### PR TITLE
Histograms initial commit

### PR DESCRIPTION
**Pull request by laullon**

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Gopkg.toml example
```
Comment (vikramraman) [inline Gopkg.toml]: Let's remove the commented out example parts.

File: metrics/histogram.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
+package wavefront
```
Comment (vikramraman) [inline metrics/histogram.go]: Put this under `histograms/histogram.go` and change package to histogram.

In terms of terminology histogram is different than metrics. They are two different products that Wavefront now offers.

**Comment by vikramraman** [comment]
Will look into it more later. Just added few minor comments for now.

**Comment by vikramraman** [comment]
Also please add unit tests for your code. It'd be required prior to merging.

File: metrics/histogram.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+	tdigest "github.com/caio/go-tdigest"
+)
+
+type Histogram struct {
+	mutex              sync.Mutex
+	priorTimedBinsList []*timedBin
+	currentTimedBin    *timedBin
+
+	metricName string
```
Comment (vikramraman) [inline metrics/histogram.go]: This doesn't belong here. Let's remove the metricName, source and tags from this. This should be purely about creating and updating a histogram. The reporting part such as name etc can be handled in user code.

File: metrics/histogram.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+	return h
+}
+
+func (h *Histogram) Add(v float64) {
+	h.rotateCurrentTDigestIfNeedIt()
+	h.currentTimedBin.tdigest.Add(v)
+}
+
+func (h *Histogram) Report(sender senders.Sender) {
```
Comment (vikramraman) [inline metrics/histogram.go]: Remove this method as well.

File: metrics/histogram.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	timestamp time.Time
+}
+
+type distribution struct {
+	centroids []senders.Centroid
+	timestamp time.Time
+}
+
+func NewHistogram(metricName string, source string, tags map[string]string) *Histogram {
+	h := &Histogram{}
```
Comment (vikramraman) [inline metrics/histogram.go]: For consistency, please use the struct form :

```
&Histogram {
MaxBins: 10,
Granularity: senders.Minute,
Compression: 5,
}
```

File: metrics/histogram.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+func (h *Histogram) Report(sender senders.Sender) {
+	distributions := h.getDistributions()
+	log.Printf("metric: '%s' count:'%d'", h.metricName, len(distributions))
+	hgs := map[senders.HistogramGranularity]bool{h.Granularity: true}
+	for _, distribution := range distributions {
+		sender.SendDistribution(h.metricName, distribution.centroids, hgs, distribution.timestamp.Unix(), h.source, h.tags)
+	}
+}
+
+func (h *Histogram) getDistributions() []distribution {
```
Comment (vikramraman) [inline metrics/histogram.go]: rename to `distributions()`. It's idiomatic in go to omit the get prefixes getters: https://golang.org/doc/effective_go.html#Getters

File: histograms/histogram.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import (
+	"sync"
+	"time"
+
+	"github.com/wavefronthq/wavefront-sdk-go/senders"
+
+	tdigest "github.com/caio/go-tdigest"
+)
+
+type Histogram struct {
```
Comment (vikramraman) [inline histograms/histogram.go]: We should define an interface named `Histogram` with the relevant exported methods and have this struct just implement that type. Can rename this struct to something else.

File: histograms/histogram.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	mutex              sync.Mutex
+	priorTimedBinsList []*timedBin
+	currentTimedBin    *timedBin
+
+	Granularity time.Duration
```
Comment (vikramraman) [inline histograms/histogram.go]: Would be better to keep the struct immutable by not exporting these variables. Let's make them all lower case and just have the `NewHistogram` constructor take these as input and then construct the histogram struct internally and return an interface.

File: histograms/histogram.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	tdigest   *tdigest.TDigest
+	timestamp time.Time
+}
+
+type Distribution struct {
+	Centroids []senders.Centroid
+	Timestamp time.Time
+}
+
+func NewHistogram() *Histogram {
```
Comment (vikramraman) [inline histograms/histogram.go]: As commented above, take relevant values as input and return the `Histogram` interface.

Comment (vikramraman) [reply histograms/histogram.go]: Refer to https://github.com/rcrowley/go-metrics/blob/master/histogram.go for example.

File: histograms/histogram.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	h := &Histogram{MaxBins: 10, Granularity: time.Minute, Compression: 5}
+	return h
+}
+
+func (h *Histogram) Add(v float64) {
```
Comment (vikramraman) [inline histograms/histogram.go]: Rename to `Update`.

File: histograms/histogram.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	MaxBins     int
+}
+
+type timedBin struct {
+	tdigest   *tdigest.TDigest
+	timestamp time.Time
+}
+
+type Distribution struct {
+	Centroids []senders.Centroid
```
Comment (vikramraman) [inline histograms/histogram.go]: Let's move all histogram related types into this `histogram` package from the `senders` package. Put them under a `types.go` file.

**Comment by vikramraman** [comment]
Overall the histo implementation LGTM. We need to expose methods to get the histogram values though:
`Min()`, `Max()`, `Mean()`, `Count()`, `Sum()`, `Quantile(double quantile)` etc. And also `Snapshot()` which will return a readonly copy of the histogram.

**Comment by laullon** [comment]
We already have a `Snapshot()`, it's called `Distributions()` it give you all the info on the T-Digest, is not read-only, but the modifications will not affect to the histogram.

File: histogram/histogram.go [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+	max := math.SmallestNonzeroFloat64
+	h.currentTimedBin.tdigest.ForEachCentroid(func(mean float64, count uint32) bool {
+		max = math.Max(max, mean)
+		return true
+	})
+	return max
+}
+
+// Max returns the minimun Value of samples on this histogram.
```
Comment (vikramraman) [inline histogram/histogram.go]: typo: Min

File: histogram/histogram.go [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+	min := math.MaxFloat64
+	h.currentTimedBin.tdigest.ForEachCentroid(func(mean float64, count uint32) bool {
+		min = math.Min(min, mean)
+		return true
+	})
+	return min
+}
+
+// Max returns the sum of all values on this histogram.
```
Comment (vikramraman) [inline histogram/histogram.go]: typo: Sum

**Comment by vikramraman** [comment]
@laullon Just some minor typo's in comments. Please fix them.

it'd also be good to add some unit tests for the min, max, mean etc. But I'm okay with adding those in a following checkin.

**Comment by marcosrmendezthd** [comment]
hi! are we doing a release with this new functionality? using go.mod i'm stuck on v0.9.1 😄

**Pull request merged**